opentelemetry-api/tests/trace/test_immutablespancontext.py

```diff
@@ -0,0 +1,61 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from opentelemetry import trace
+from opentelemetry.trace import TraceFlags, TraceState
+
+
+class TestImmutableSpanContext(unittest.TestCase):
+    def test_ctor(self):
+        context = trace.SpanContext(
+            1,
+            1,
+            is_remote=False,
+            trace_flags=trace.DEFAULT_TRACE_OPTIONS,
+            trace_state=trace.DEFAULT_TRACE_STATE,
+        )
+
+        self.assertEqual(context.trace_id, 1)
+        self.assertEqual(context.span_id, 1)
+        self.assertEqual(context.is_remote, False)
+        self.assertEqual(context.trace_flags, trace.DEFAULT_TRACE_OPTIONS)
+        self.assertEqual(context.trace_state, trace.DEFAULT_TRACE_STATE)
+
+    def test_attempt_change_attributes(self):
+        context = trace.SpanContext(
+            1,
+            2,
+            is_remote=False,
+            trace_flags=trace.DEFAULT_TRACE_OPTIONS,
+            trace_state=trace.DEFAULT_TRACE_STATE,
+        )
+
+        with self.assertRaises(TypeError):
+            context.trace_id = 2
+        with self.assertRaises(TypeError):
+            context.span_id = 3
+        with self.assertRaises(TypeError):
+            context.is_remote = True
+        with self.assertRaises(TypeError):
+            context.trace_flags = TraceFlags(3)
+        with self.assertRaises(TypeError):
+            context.trace_state = TraceState([("test", "test")])
```

Comment:
`mypy` produces an error for these lines with `error: Property "trace_state" defined in "SpanContext" is read-only` (https://github.com/open-telemetry/opentelemetry-python/runs/1130417833?check_suite_focus=true). 

However, I think these tests are useful in seeing if each attribute in SpanContext is actually immutable. Is there a way to relax the restrictions on this test file or are these tests unnecessary? Thank you!

Reply:
I agree these tests are important. How does mypy actually know that trace_state is read-only? Is it something about the class (maybe because it's a tuple)? Regardless, maybe there's a way to ignore this error inline?

Reply:
Hmm, I think `mypy` actually checks the annotations of the properties. After some investigation, it seems that if only `@property` is set (and not `@property.setter`), then `mypy` determines the attribute to be read-only. 

I wasn't able to find a way to ignore this error inline. It's possible to make an empty setter for each of the attributes (but that seems a little clunky). Another solution is to comment out `check_untyped_defs` in the `mypy-relaxed.ini` file.